modules/layers/src/arc-layer/arc-layer.js

```diff
@@ -143,18 +143,18 @@ export default class ArcLayer extends Layer {
   }
 
   calculateInstancePositions(attribute) {
-    const {data, getSourcePosition, getTargetPosition} = this.props;
+    const {getSourcePosition, getTargetPosition} = this.props;
     const {value, size} = attribute;
     let i = 0;
-    for (const object of data) {
-      const sourcePosition = getSourcePosition(object);
-      const targetPosition = getTargetPosition(object);
+    this.iterateData((object, context) => {
```

Comment:
Nit: `context` is not a good name, it is overloaded and empty of meaning at the same time. There may not be any better choice, but let's brainstorm and try.

Reply:
What I am not completely happy about here is that this system leaves the burden of handling binary data (which should be the primary, most heavily optimized case, to the app).